ffi-convert/src/types.rs

```diff
@@ -114,38 +116,60 @@ pub struct CArray<T> {
     size: usize,
 }
 
-impl<U: AsRust<V>, V> AsRust<Vec<V>> for CArray<U> {
+impl<U: AsRust<V> + 'static, V> AsRust<Vec<V>> for CArray<U> {
     fn as_rust(&self) -> Result<Vec<V>, AsRustError> {
         let mut vec = Vec::with_capacity(self.size);
+
+        let values = unsafe { std::slice::from_raw_parts_mut(self.data_ptr as *mut U, self.size) };
+
         if self.size > 0 {
-            let values =
-                unsafe { std::slice::from_raw_parts_mut(self.data_ptr as *mut U, self.size) };
-            for value in values {
-                vec.push(value.as_rust()?);
+            if is_primitive(TypeId::of::<U>()) {
+                unsafe {
+                    ptr::copy(
+                        values.as_ptr() as *const V,
+                        vec.as_mut_ptr() as *mut V,
+                        self.size,
+                    );
+                    vec.set_len(self.size);
+                }
+            } else {
+                for value in values {
+                    vec.push(value.as_rust()?);
+                }
             }
         }
         Ok(vec)
     }
 }
 
-impl<U: CReprOf<V> + CDrop, V> CReprOf<Vec<V>> for CArray<U> {
+impl<U: CReprOf<V> + CDrop, V: 'static> CReprOf<Vec<V>> for CArray<U> {
     fn c_repr_of(input: Vec<V>) -> Result<Self, CReprOfError> {
         let input_size = input.len();
-        Ok(Self {
-            data_ptr: if input_size > 0 {
-                Box::into_raw(
+        let mut output: CArray<U> = CArray {
+            data_ptr: ptr::null(),
+            size: input_size,
+        };
+
+        if input_size > 0 {
+            if is_primitive(TypeId::of::<V>()) {
+                let mut data: MaybeUninit<Vec<V>> = unsafe { MaybeUninit::uninit().assume_init() };
+                data.write(input);
+                let data = unsafe { mem::transmute::<_, Vec<V>>(data) };
+                output.data_ptr = Box::into_raw(data.into_boxed_slice()) as *const U;
```

Comment:
```suggestion
                output.data_ptr = Box::into_raw(input.into_boxed_slice()) as *const U;
```
this should do the trick directly without having to do nasty stuff with maybe uninit and transmute

Reply:
Ok, it's simpler